feat: NTRN-332. Add tokenfactory module (#131)
neutron-org/neutron: x/tokenfactory/README.md

# Token Factory (by Osmosis Labs)

> > This module was taken from Osmosis chain codebase (commit: https://github.com/osmosis-labs/osmosis/commit/9e178a631f91ffc91c51f3665ed915c9f15e1807). The reason of this action was to adopt module and tests to our codebase because it was not possible to import it without code modification
> > that was made by Osmosis team to the orginal Cosmos SDK. These changes made it not possible (without deep modifications of the whole code) to import module to our code.
> > Also support of the creation fee was removed at the moment because we do not have community pool in the Neutron. 

Reviewer: I believe this version is much easier to read:
```suggestion
> This module is vendored from Osmosis chain codebase (commit: https://github.com/osmosis-labs/osmosis/commit/9e178a631f91ffc91c51f3665ed915c9f15e1807) and adopted for our codebase (including unit tests). It was impossible to import it directly without patches to original Cosmos SDK by Osmosis team, which we don't have.
> Support for creation fee is currently removed, since we don't have community pool in Neutron. 
```

message MsgCreateDenom {
  string sender = 1 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
  string subdenom = 2 [ (gogoproto.moretags) = "yaml:\"subdenom\"" ];
}
```

**State Modifications:**
- Fund community pool with the denom creation fee from the creator address, set in `Params`
- Set `DenomMetaData` via bank keeper
- Set `AuthorityMetadata` for the given denom to store the admin for the created denom `factory/{creator address}/{subdenom}`. Admin is automatically set as the Msg sender
- Add denom to the `CreatorPrefixStore`, where a state of denoms created per creator is kept

### Mint
- Minting of a specific denom is only allowed for the creator of the denom registered during `CreateDenom`
``` {.go}
message MsgMint {
  string sender = 1 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
  cosmos.base.v1beta1.Coin amount = 2 [
    (gogoproto.moretags) = "yaml:\"amount\"",
    (gogoproto.nullable) = false
  ];
}
```

**State Modifications:**
- Safety check the following
  - Check that the denom minting is created via `tokenfactory` module
  - Check that the sender of the message is the admin of the denom
- Mint designated amount of tokens for the denom via `bank` module

### Burn

> - Burning of a specific denom is only allowed for the creator of the denom registered during `CreateDenom`

Reviewer: I believe I didn't understand this sentence correctly, so please take my suggestion with caution. In case if I am really wrong, please rewrite this sentence anyway. It is pretty confusing.
```suggestion
- Burning of a specific denom is only allowed for the creator of the denom registered with `CreateDenom`
```

``` {.go}
message MsgBurn {
  string sender = 1 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
  cosmos.base.v1beta1.Coin amount = 2 [
    (gogoproto.moretags) = "yaml:\"amount\"",
    (gogoproto.nullable) = false
  ];
}
```

**State Modifications:**

> - Saftey check the following

Reviewer: ```suggestion
- Safety check the following
```

`factory/{creator address}/{subdenom}`
Splitting up into sub-components, this has:
* `len(factory) = 7`
* `2 * len("/") = 2`
* `len(longest_subdenom)`
* `len(creator_address) = len(bech32(longest_addr_length, chain_addr_prefix))`.
Longest addr length at the moment is `32 bytes`.
Due to SDK error correction settings, this means `len(bech32(32, chain_addr_prefix)) = len(chain_addr_prefix) + 1 + 58`.
Adding this all, we have a total length constraint of `128 = 7 + 2 + len(longest_subdenom) + len(longest_chain_addr_prefix) + 1 + 58`.
Therefore `len(longest_subdenom) + len(longest_chain_addr_prefix) = 128 - (7 + 2 + 1 + 58) = 60`.

The choice between how we standardized the split these 60 bytes between maxes from longest_subdenom and longest_chain_addr_prefix is somewhat arbitrary. Considerations going into this:
* Per [BIP-0173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32) the technically longest HRP for a 32 byte address ('data field') is 31 bytes. (Comes from encode(data) = 59 bytes, and max length = 90 bytes)
* subdenom should be at least 32 bytes so hashes can go into it
* longer subdenoms are very helpful for creating human readable denoms

> * chain addresses should prefer being smaller. The longest HRP in cosmos to date is 11 bytes. (`persistence`)

Reviewer: ```suggestion
* it should be preferred to use short chain addresses. The longest HRP in cosmos to date is 11 bytes. (`persistence`)
```

For explicitness, its currently set to `len(longest_subdenom) = 44` and `len(longest_chain_addr_prefix) = 16`.

> Please note, if the SDK increases the maximum length of a denom from 128 bytes, these caps should increase.

Reviewer: ```suggestion
Please note, if the SDK ever increases maximum denom length from 128 bytes, these caps will increase accordingly.
```
